### PR TITLE
Improve router logging and update documentation

### LINKED ISSUE #1646: Improve logging of pathfinding behavior
The behavior of sending a payment can be quite confusing to users, in particular if an otherwise obvious channel choice is avoided without giving any feedback as to why. 

We therefore should try to improve the logging in `get_route` without getting too spammy. In particular, we should find a way to give users feedback when for example a direct channel is avoided due to lacking channel reserves, its value contribution, or similar.

### PR DESCRIPTION
**Pull request by tnull**
Fixes #1646 

As #2417 will slip to a future release, this PR just includes the improved logging and documentation fixups of of our router.

Previously, we barely gave any hints why we excluded certain hops during pathfinding. Here, we introduce more verbose logging by a) accounting how much candidates we ignored for which reasons and b) logging any first/last/blinded hops we end up ignoring.

We also improve documentation as `RouteParameters` are not included in `Event::PaymentPathFailed` since 0.0.115, and we don't give value/payee as immediate arguments to `find_route` anymore.

### CODE REVIEW

File: lightning/src/routing/router.rs [hunk 1]
```diff
@@ -1714,13 +1728,37 @@ where L::Target: Logger {
...
-						exceeds_cltv_delta_limit || payment_failed_on_this_channel {
-						// Path isn't useful, ignore it and move on.
+					if !contributes_sufficient_value {
+						if should_log_candidate {
+							log_trace!(logger, "Ignoring {} due to insufficient value contribution.", LoggedCandidateHop(&$candidate));
+						}
+						num_ignored_value_contribution += 1;
+					} else if exceeds_max_path_length {
+						if should_log_candidate {
+							log_trace!(logger, "Ignoring {} due to exceeding max. path length limit.", LoggedCandidateHop(&$candidate));
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:1749]: nit: no need to use a `.` when abbreviating `max` IMO

File: lightning/src/routing/router.rs [hunk 2]
```diff
@@ -2322,6 +2360,12 @@ where L::Target: Logger {
 		}
 	}
 
+	let num_ignored_total = num_ignored_value_contribution + num_ignored_path_length_limit +
+		num_ignored_cltv_delta_limit + num_ignored_previously_failed;
+	if num_ignored_total > 0 {
+		log_trace!(logger, "Ignored {} candidate hops due to insufficient value contrib., {} due to path length limit, {} due to CLTV delta limit, {} due to previous payment failure. Total: {} ignored candidates.", num_ignored_value_contribution, num_ignored_path_length_limit, num_ignored_cltv_delta_limit, num_ignored_previously_failed, num_ignored_total);
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:2366]: Likewise for `contrib.` or just spell it out as done earlier for consistency when searching the log.

File: lightning/src/routing/router.rs [hunk 3]
```diff
@@ -1714,13 +1728,37 @@ where L::Target: Logger {
...
+					} else if exceeds_cltv_delta_limit {
+						if should_log_candidate {
+							log_trace!(logger, "Ignoring {} due to exceeding CLTV delta limit.", LoggedCandidateHop(&$candidate));
+						}
+						num_ignored_cltv_delta_limit += 1;
+					} else if payment_failed_on_this_channel {
+						if should_log_candidate {
+							log_trace!(logger, "Ignoring {} due to a failed previous payment attempt.", LoggedCandidateHop(&$candidate));
+						}
+						num_ignored_previously_failed += 1;
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:1761]: Are these actually mutually exclusive?

Comment (valentinewallace) [reply lightning/src/routing/router.rs:1761]: True, might be best to just have one long log that includes all the variables

**Review by jkczyz — COMMENTED** [review]
LGTM. Please squash

**Review by jkczyz — APPROVED** [review]

**Pull request merged**